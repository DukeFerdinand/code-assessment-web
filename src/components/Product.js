import React from 'react'
import PropTypes from 'prop-types'
import ReactSVG from 'react-svg'

import images from '../assets/images'
import icons from '../assets/icons/index'

const Product = ({ price, inventory, title, quantity, inCart, removeProductFromCart }) => (
  <div className="product">
    {
      inCart ? (
        // Again as before, these images should be fed from a Database,
        // but that's outside the scope of this project
        <span className="product-image-wrapper">
          <img alt={title} className="product-image" src={images.thumbnails[`${title}Thumb`] || images.NotFound} />
        </span>
      ) : null
    }
    <div className="product-info-upper">
      <div className="product-title">
        {title}
      </div>
      <div className="product-price">
        &#36;{price}
      </div>
      { inCart ? <span onClick={() => removeProductFromCart()} className="remove-product-link">Remove</span> : null }
    </div>
    {
      !inCart ? (
        <div className={`product-availability ${inventory ? '' : 'out-of-stock'}`}>
          {inventory ? `${inventory} REMAINING` : 'OUT OF STOCK'}
        </div>
      ) : (
        <div className="product-quantity">
          <button
            className="minus-button"
            onClick={() => removeProductFromCart(1)}
            >
            <ReactSVG
              className="button-icon-wrapper"
              svgClassName="button-icon"
              src={icons.minus}
            />
          </button>
          <span className="product-quantity-input-wrapper">
            <input className="product-quantity-input" type="text" value={quantity}/>
          </span>
          <button className="plus-button">
            <ReactSVG
              className="button-icon-wrapper"
              svgClassName="button-icon"
              src={icons.plus}
            />
          </button>
        </div>
      )
    }
  </div>
)

Product.propTypes = {
  price: PropTypes.number,
  inventory: PropTypes.number,
  title: PropTypes.string,
  quantity: PropTypes.number,
  inCart: PropTypes.bool,
  removeProductFromCart: PropTypes.func
}

export default Product
