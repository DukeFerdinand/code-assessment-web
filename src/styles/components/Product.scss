@import '../shared';

.product-wrapper {
  width: 100%;
  height: auto;
  max-height: 275px;
  display: flex;
  margin: 30px auto;

  @media (max-width: $tablet) {
    max-height: 175px;
    margin: 20px auto;
  }
  .product-image {
    flex: .8;
    background: black + 200;
    height: 100%;
    max-height: 250px;
    max-width: 400px;
    border-radius: 10px 0 0 10px;

    @media (max-width: $tablet) {
      max-width: 272px;
      max-height: 175px;
      border-radius: 5px 0 0 5px;
    }
  }
  .product-info-wrapper {
    flex: 1;
    border-radius: 0 8px 8px 0;
    background: white;
    .product-info {
      box-sizing: border-box;
      height: 100%;
      padding: 30px 60px;
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      @media (max-width: $tablet) {
        padding: 20px 20px 20px 40px;
      }

      .product-info-upper {
        display: flex;
        justify-content: space-between;

        .product-title {
          font-size: 2.5em;
          color: $text-color-main;

          @media (max-width: $tablet) {
            font-size: 1.5em;
          }
        }
        .product-price {
          font-size: 1.3em;
          position: relative;
          top: 10px;

          @media (max-width: $tablet) {
            font-size: 1.1em;
            top: 5px;
          }
        }
      }
      .product-availability {
        margin: 10px 5px;
        font-size: .90em;
        color: $text-color-secondary;
        &.out-of-stock {
          color: #e74c3c;
        }

        @media (max-width: $tablet) {
          font-size: .75em;
        }
      }
      .product-action {
        width: 130px;
        background: $color-primary;
        border-radius: 20px;
        color: white;
        cursor: pointer;
        font-size: .75em;
        padding: 10px 10px;

        @media (max-width: $tablet) {
          bottom: 0px;
          padding: 8px;
          width: 120px;
          font-size: .7em;
        }

        &:hover, &:focus {
          background: $button-primary-hover
        }

        &:disabled {
          cursor: initial;
          background: $color-primary-disabled;
        }
      }
    }
  }
}

.cart-content {
  .product {
    display: flex;
    // background: ;
    flex-wrap: wrap;
    margin-bottom: 25px;
    .product-image-wrapper {
      width: 150px;
      // height: 125px;
      .product-image {
        width: 100%;
      }
    }
    .product-info-upper {
      height: 100px;
      margin-left: 10px;
      position: relative;
      .product-title {
        font-size: 1.25em;
      }
      .product-price {
        margin-top: 5px;
      }
      .remove-product-link {
        // TODO: switch back to cursor
        cursor: not-allowed;
        flex: .01;
        color: red - 10;
        position: absolute;
        bottom: 10px;
      }
    }
    .product-quantity {
      display: flex;
      margin: 10px auto 0 auto;
      border-radius: 60px;

      %quantity-button {
        //TODO: Re-implement this once the cart buttons work
        // cursor: pointer;
        cursor: not-allowed;
        height: 40px;
        width: 80px;
        flex: 1;
        background: $button-secondary;
        // TODO: Re-implement once cart buttons work
        // &:hover {
        //   background: $button-primary;
        //   * {
        //     fill: white;
        //   }
        // }
      }
      .minus-button {
        @extend %quantity-button;
        border-radius: 30px 0 0 30px;
        .button-icon-wrapper {
          margin: 0 auto 5px auto;
          width: 15px;
          .button-icon {
            width: 15px;
          }
        }
      }
      .product-quantity-input-wrapper {
        background: $button-secondary;
        height: 40px;
        flex: 1;
        .product-quantity-input {
          border: none;
          font: $app-font;
          font-size: 1em;
          margin-top: 2px;
          height: 34px;
          width: 100%;
          text-align: center;
        }
      }

      .plus-button {
        @extend %quantity-button;
        border-radius: 0 30px 30px 0;

        .button-icon-wrapper {
          margin: 2px auto 0 auto;
          padding-right: 10px;
          width: 15px;
          .button-icon {
            width: 15px;
          }
        }
      }
    }
  }
}