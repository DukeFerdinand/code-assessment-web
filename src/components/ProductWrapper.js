import React from 'react'
import PropTypes from 'prop-types'
import Product from './Product'

import '../styles/components/Product.scss'

const ProductWrapper = ({ product, onAddToCartClicked }) => (
  <div className="product-wrapper">

    {/* This would ideally be replaced with a DB stored url */}
    <img className="product-image" src={`../assets/${product.title}.png`} alt="test"/>

    <div className="product-info-wrapper">
      <div className="product-info">
        <Product
          title={product.title}
          price={product.price}
          inventory={product.inventory} />
        <button
          className="product-action"
          onClick={onAddToCartClicked}
          disabled={product.inventory > 0 ? '' : 'disabled'}>
          {product.inventory > 0 ? 'ADD TO CART' : 'SOLD OUT'}
        </button>
      </div>
    </div>
  </div>
)

ProductWrapper.propTypes = {
  product: PropTypes.shape({
    title: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    inventory: PropTypes.number.isRequired
  }).isRequired,
  onAddToCartClicked: PropTypes.func.isRequired
}

export default ProductWrapper
