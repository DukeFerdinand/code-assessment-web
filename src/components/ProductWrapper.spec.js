import React from 'react'
import { shallow } from 'enzyme'
import Product from './Product'
import ProductWrapper from './ProductWrapper'

const setup = product => {
  const actions = {
    onAddToCartClicked: jest.fn()
  }

  const component = shallow(
    <ProductWrapper product={product} {...actions} />
  )

  return {
    component: component,
    actions: actions,
    button: component.find('button'),
    product: component.find(Product)
  }
}

let productProps

describe('ProductWrapper component', () => {
  beforeEach(() => {
    productProps = {
      title: 'Product 1',
      price: 9.99,
      inventory: 6
    }
  })

  it('should render product', () => {
    const { product } = setup(productProps)
    expect(product.props()).toEqual({ title: 'Product 1', price: 9.99, inventory: 6 })
  })

  it('should render Add To Cart message', () => {
    const { button } = setup(productProps)
    expect(button.text()).toMatch(/^ADD TO CART/)
  })

  it('should not disable button', () => {
    const { button } = setup(productProps)
    expect(button.prop('disabled')).toEqual('')
  })

  it('should call action on button click', () => {
    const { button, actions } = setup(productProps)
    button.simulate('click')
    expect(actions.onAddToCartClicked).toBeCalled()
  })

  describe('when product inventory is 0', () => {
    beforeEach(() => {
      productProps.inventory = 0
    })

    it('should render Sold Out message', () => {
      const { button } = setup(productProps)
      expect(button.text()).toMatch(/^SOLD OUT/)
    })

    it('should disable button', () => {
      const { button } = setup(productProps)
      expect(button.prop('disabled')).toEqual('disabled')
    })
  })
})
