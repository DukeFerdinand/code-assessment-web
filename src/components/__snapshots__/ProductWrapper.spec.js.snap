// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductWrapper component should match snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="product-wrapper"
  >
    <img
      alt="test"
      className="product-image"
      src="NotFound.jpg"
    />
    <div
      className="product-info-wrapper"
    >
      <div
        className="product-info"
      >
        <Product
          inventory={6}
          price={9.99}
          title="Product 1"
        />
        <button
          className="product-action"
          disabled=""
          onClick={[MockFunction]}
        >
          ADD TO CART
        </button>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div
      className="product-wrapper"
    >
      <img
        alt="test"
        className="product-image"
        src="NotFound.jpg"
      />
      <div
        className="product-info-wrapper"
      >
        <div
          className="product-info"
        >
          <Product
            inventory={6}
            price={9.99}
            title="Product 1"
          />
          <button
            className="product-action"
            disabled=""
            onClick={[MockFunction]}
          >
            ADD TO CART
          </button>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ProductWrapper
        onAddToCartClicked={[MockFunction]}
        product={
          Object {
            "inventory": 6,
            "price": 9.99,
            "title": "Product 1",
          }
        }
      />,
      "_debugID": 9,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "onAddToCartClicked": [MockFunction],
          "product": Object {
            "inventory": 6,
            "price": 9.99,
            "title": "Product 1",
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 5,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="product-wrapper"
        >
          <img
            alt="test"
            className="product-image"
            src="NotFound.jpg"
          />
          <div
            className="product-info-wrapper"
          >
            <div
              className="product-info"
            >
              <Product
                inventory={6}
                price={9.99}
                title="Product 1"
              />
              <button
                className="product-action"
                disabled=""
                onClick={[MockFunction]}
              >
                ADD TO CART
              </button>
            </div>
          </div>
        </div>,
        "_debugID": 10,
        "_renderedOutput": <div
          className="product-wrapper"
        >
          <img
            alt="test"
            className="product-image"
            src="NotFound.jpg"
          />
          <div
            className="product-info-wrapper"
          >
            <div
              className="product-info"
            >
              <Product
                inventory={6}
                price={9.99}
                title="Product 1"
              />
              <button
                className="product-action"
                disabled=""
                onClick={[MockFunction]}
              >
                ADD TO CART
              </button>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ProductWrapper
    onAddToCartClicked={[MockFunction]}
    product={
      Object {
        "inventory": 6,
        "price": 9.99,
        "title": "Product 1",
      }
    }
  />,
}
`;
